File: 2022/07.py
import re


def get_input(puzzle: str) -> list[str]:
    return puzzle.splitlines()


def get_dir_sizes(lines: list[str]) -> dict[tuple[str, ...], int]:
    out = {}
    seen = set()
    pwd = ()
    for line in lines:
        if match := re.match(r"^\$ cd (.+)$", line):
            d = match[1]
            if d == "/":
                pwd = ()
            elif d == "..":
                pwd = pwd[:-1]
            else:
                pwd = (*pwd, d)
        elif match := re.match(r"^(\d+) (.+)$", line):
            s = int(match[1])
            n = match[2]

            if (pwd, n) in seen:
                continue
            seen.add((pwd, n))

            d = pwd
            while True:
                out[d] = out.get(d, 0) + s
                if not d:
                    break
                d = d[:-1]

    return out


def part1(puzzle: str):
    sizes = get_dir_sizes(get_input(puzzle))
    return sum(s for s in sizes.values() if s <= 100000)


def part2(puzzle: str):
    sizes = get_dir_sizes(get_input(puzzle))
    free = 70000000 - sizes[()]
    return min(s for s in sizes.values() if s >= 30000000 - free)


if __name__ == "__main__":
    from aoc import run

    run(2022, 7, part1, part2)

File: 2022/07.rs
use std::collections::{HashMap, HashSet};

type Input = (u64, Vec<u64>);

fn setup(input: &str) -> Input {
    let mut out = HashMap::new();
    let mut seen = HashSet::new();
    let mut pwd = vec![];
    for line in input.trim().lines() {
        let mut split = line.split(' ');
        match (split.next(), split.next(), split.next()) {
            (Some("$"), Some("cd"), Some(dir)) => {
                if dir == "/" {
                    pwd.clear();
                } else if dir == ".." {
                    pwd.pop();
                } else {
                    pwd.push(dir);
                }
            }
            (Some(size), Some(name), None) => {
                if let Ok(size) = size.parse::<u64>() {
                    let key = (pwd.join("/"), name);
                    if seen.contains(&key) {
                        continue;
                    }
                    seen.insert(key);

                    for i in 0..=pwd.len() {
                        out.entry(pwd[..i].join("/"))
                            .and_modify(|x| *x += size)
                            .or_insert(size);
                    }
                }
            }
            _ => {}
        }
    }
    (out[""], out.values().copied().collect())
}

fn part1((_, sizes): &Input) -> u64 {
    sizes.iter().filter(|x| **x <= 100000).sum()
}

fn part2((root_size, sizes): &Input) -> u64 {
    let free = 70000000 - root_size;
    *sizes
        .iter()
        .filter(|x| **x >= 30000000 - free)
        .min()
        .unwrap()
}

aoc::main!(2022, 7);
aoc::example!(ex01, "07.1.txt", 95437, 24933642);
